### ISSUE TITLE
Unsupervised GraphSage

### ISSUE THREAD

**Issue by Sharad24** [pyg-team/pytorch_geometric#1420]
## ❓ Questions & Help

Is there a complete example of the Unsupervised GraphSage with its loss formulation?

**Comment by rusty1s** [comment]
We don't have an example for this yet. However, implementing it should be straightforward via `torch_geometric.utils.negative_sampling`.

**Comment by Sharad24** [comment]
Is it something thats planned to be added? If yes, I could work on it.

**Comment by rusty1s** [comment]
It would be awesome if you would contribute that:)

**Comment by arunavsk** [comment]
@Sharad24  I think it would be a lot similar to https://github.com/rusty1s/pytorch_geometric/blob/master/examples/autoencoder.py , but using `SAGEConv` and `NeighborSampler`

@rusty1s thoughts?

**Comment by rusty1s** [comment]
That's true. It's a combination of `examples/reddit.py` and `examples/autoencoder.py`. One can start with the `reddit.py` example and replace its loss formulation. Note that you need two forward passes now, one for positive edges and one for negative ones.

**Comment by Sharad24** [comment]
Yup, thanks! I'll raise a PR when done

**Comment by arunavsk** [comment]
@rusty1s so I did an implementation of this using `examples/reddit.py` and `examples/autoencoder.py` https://gist.github.com/arunavsk/7a4091ccddcbfa6eb31c35c5ce7fe462
I used `recon_loss()` from `nn.GAE` which I guess in turn uses the positive edges and `utils.negative_sampling()` to calculate the total loss.

**Comment by rusty1s** [comment]
Cool, feel free to send a PR :)

**Comment by rusty1s** [comment]
Examples are now provided in `examples/` :)

**Comment by aufacci** [comment]
Hi, can you tell me which files in `examples/`  I can find the implementation of the graph-based loss function ?

Thank you @rusty1s !

**Comment by rusty1s** [comment]
Ups, please see `examples/graph_sage_unsup.py` and `examples/graph_sage_unsup_ppi.py` :)

**Comment by forrestw2005** [comment]
> 

Hey, do these codes use the loss function mentioned in the original paper? here is the loss function: 
![image](https://github.com/pyg-team/pytorch_geometric/assets/32083401/b4cf5296-1bc4-468b-84ef-ae1723a4b558)?

**Comment by rusty1s** [comment]
Yes, this is correct.

**Comment by andytwigg** [comment]
Hi, sorry for my ignorance but I'm struggling to see how `examples/graph_sage_unsup_ppi.py` implements the random walk loss function. My understanding is that `LinkNeighborLoader` returns both positive and negative edge samples, and then we take inner product of node embeddings + crossentropy loss. Doesn't this end up computing the "link prediction" loss, not the "random walk" loss. Did I miss something? Thanks!

**Comment by rusty1s** [comment]
Oh, so you are talking about the random walk based negative sampling procedure. Yes, it is correct that this is not automatically performed by `LinkNeighborLoader`.
